node/bindnode/infer.go

```diff
@@ -430,8 +433,16 @@ func inferSchema(typ reflect.Type, level int) schema.Type {
 		typSchema := schema.SpawnList(name, etypSchema.Name(), nullable)
 		defaultTypeSystem.Accumulate(typSchema)
 		return typSchema
+	case reflect.Interface:
+		if typ == goTypeLink {
+			return schemaTypeLink
+		}
+		if typ == goTypeNode {
```

Comment:
Why yes there is .. thanks for thinking about this. I've replaced this with an `AssignableTo()` which is the check we're looking for here. That _should_ work in node.go when we're operating on the type because we're doing assignments. I'm not adding any extra tests for this though, there's just too many paths to pass a custom type to in order to check this, so we may find surprises. When I get to #338 I'm going to have to add additional tests to a bunch of places to expand `Any`'s possible types anyway so maybe I'll get to it then. I'll put a note there.

I've given the same treatment to `goTypeLink` and above in the struct section - so "is struct but is assignable to `datamodel.Link` or `cid.Cid`"—where it matters later on there's a type switch which matches `datamodel.Link`, so anything that implements that interface should be caught there. There's also assignment that happens. I think we're pretty safe with these.

Reply:
actually .. it may be more complicated than that, the assignability needs to go the other way too, an `AssignX` needs to be able to take something implementing `datamodel.Node` and assign it to a field, so it's not good enough just to be able to _read_ that field as a `datamodel.Node`, we also need to be able to set it, and at the moment we're not doing any fancy type conversion or wrapping, in that position at least, it's a plain `Set()`

will consider this further later, blocked for now

Reply:
I've reverted the latest change and gone back to exact matching of link types and `datamodel.Node` rather than looking at assignability - because of the need to both get and set, I don't think we can have that much flexibility here. At least, such flexibility should be a TODO if someone wants it in the future.